kroxylicious/src/main/java/io/kroxylicious/proxy/internal/FilterHandler.java

```diff
@@ -38,13 +42,17 @@ public class FilterHandler extends ChannelDuplexHandler {
     private final KrpcFilter filter;
     private final long timeoutMs;
     private final String sniHostname;
+    private final Channel inboundChannel;
     private final FilterInvoker invoker;
+    private CompletableFuture<Void> writeFuture = CompletableFuture.completedFuture(null);
```

Comment:
How do you feel about queuing up Futures and chaining them together, as opposed to having something here like a `Deque<DelayedRequest>` and `Deque<DelayedResponse>` plus variables to track what state we are in, like `boolean requestDeferred`.

Using Future lets us combine those states.

Reply:
I'm okay with that approach.